File: android/hermes-livechat/src/main/java/com/mobene/hermes/livechat/internal/ApiClient.kt
package com.mobene.hermes.livechat.internal

import com.mobene.hermes.livechat.Conversation
import com.mobene.hermes.livechat.HermesLiveChatConfig
import com.mobene.hermes.livechat.HermesLiveChatError
import com.mobene.hermes.livechat.HermesLiveChatException
import com.mobene.hermes.livechat.Message
import com.mobene.hermes.livechat.SendMessageResult
import com.mobene.hermes.livechat.VisitorIdentity
import com.mobene.hermes.livechat.normalizedBaseUrl
import java.util.concurrent.TimeUnit
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONArray
import org.json.JSONObject

internal class ApiClient(private val config: HermesLiveChatConfig) {
    private val jsonMedia = "application/json; charset=utf-8".toMediaType()
    private val http = OkHttpClient.Builder()
        .callTimeout(config.requestTimeoutMillis, TimeUnit.MILLISECONDS)
        .build()

    suspend fun publicConfig(locale: String?): JSONObject = get(
        "/api/livechat/v1/public-config",
        mapOf(
            "channel_type" to "app",
            "app_key" to config.appKey,
        ) + listOfNotNull(locale?.let { "locale" to it }).toMap(),
    )

    suspend fun init(identity: VisitorIdentity, oldToken: String?): JSONObject {
        val user = JSONObject().apply {
            putOpt("email", identity.email)
            putOpt("name", identity.name)
            putOpt("avatar", identity.avatar)
        }
        val body = JSONObject().apply {
            put("channel_type", "app")
            put("app_key", config.appKey)
            putOpt("customer_id", identity.customerId)
            putOpt("external_user_id", identity.externalUserId)
            putOpt("business_id", identity.businessId)
            putOpt("ticket_id", identity.ticketId)
            putOpt("number", identity.number)
            put("user", user)
            putOpt("locale", identity.locale)
            putOpt("identity_token", identity.identityToken)
            if (identity.attrs != null) put("attrs", JSONObject(identity.attrs))
        }
        return post("/api/livechat/v1/init", body, oldToken)
    }

    suspend fun sendText(token: String, conversationId: String?, text: String, clientMsgId: String): SendMessageResult {
        val body = JSONObject().apply {
            putOpt("conversation_id", conversationId)
            put("client_msg_id", clientMsgId)
            put("content_type", "text")
            put("content", JSONObject(mapOf("text" to text)))
        }
        return SendMessageResult.fromJson(post("/api/livechat/v1/messages", body, token))
    }

    suspend fun sendImage(
        token: String,
        conversationId: String?,
        key: String,
        url: String,
        mimeType: String,
        size: Int,
        clientMsgId: String,
    ): SendMessageResult {
        val body = JSONObject().apply {
            putOpt("conversation_id", conversationId)
            put("client_msg_id", clientMsgId)
            put("content_type", "image")
            put("content", JSONObject(mapOf("key" to key, "url" to url, "mime" to mimeType, "size" to size)))
        }
        return SendMessageResult.fromJson(post("/api/livechat/v1/messages", body, token))
    }

    suspend fun markRead(token: String, messageId: String) {
        post("/api/livechat/v1/messages/${urlEncode(messageId)}/read", null, token)
    }

    suspend fun history(token: String, conversationId: String, afterId: String?, limit: Int): List<Message> {
        val json = get(
            "/api/livechat/v1/conversations/${urlEncode(conversationId)}/messages",
            mapOf(
                "limit" to limit.toString(),
            ) + listOfNotNull(afterId?.let { "after_id" to it }).toMap(),
            token,
        )
        val items = json.optJSONArray("items") ?: JSONArray()
        return (0 until items.length()).map { Message.fromJson(items.getJSONObject(it)) }
    }

    suspend fun conversations(token: String): List<Conversation> {
        val json = get("/api/livechat/v1/conversations", mapOf("limit" to "20"), token)
        val items = json.optJSONArray("items") ?: JSONArray()
        return (0 until items.length()).map { Conversation.fromJson(items.getJSONObject(it)) }
    }

    suspend fun presign(token: String, filename: String, mimeType: String, size: Int): JSONObject = post(
        "/api/livechat/v1/attachments/presign",
        JSONObject(mapOf("filename" to filename, "mime" to mimeType, "size" to size)),
        token,
    )

    suspend fun uploadPresigned(
        url: String,
        method: String,
        headers: Map<String, String>,
        bytes: ByteArray,
        mimeType: String,
    ) = withContext(Dispatchers.IO) {
        val body = bytes.toRequestBody(mimeType.toMediaType())
        val builder = Request.Builder().url(url).method(method, body)
        headers.forEach { (key, value) -> builder.header(key, value) }
        http.newCall(builder.build()).execute().use { resp ->
            if (!resp.isSuccessful) {
                throw HermesLiveChatException(
                    HermesLiveChatError.ATTACHMENT_TYPE_INVALID,
                    message = "attachment upload failed",
                    status = resp.code,
                )
            }
        }
    }

    private suspend fun get(path: String, query: Map<String, String>, token: String? = null): JSONObject {
        val url = config.normalizedBaseUrl() + path + "?" + query.entries.joinToString("&") {
            "${urlEncode(it.key)}=${urlEncode(it.value)}"
        }
        val request = Request.Builder().url(url).applyHeaders(token).get().build()
        return execute(request)
    }

    private suspend fun post(path: String, body: JSONObject?, token: String?): JSONObject {
        val requestBody = body?.toString()?.toRequestBody(jsonMedia)
        val request = Request.Builder()
            .url(config.normalizedBaseUrl() + path)
            .applyHeaders(token, body != null)
            .post(requestBody ?: ByteArray(0).toRequestBody(null))
            .build()
        return execute(request)
    }

    private suspend fun execute(request: Request): JSONObject = withContext(Dispatchers.IO) {
        http.newCall(request).execute().use { response ->
            val text = response.body?.string().orEmpty()
            val payload = if (text.isNotEmpty()) JSONObject(text) else JSONObject()
            val code = payload.opt("code")?.toString()?.toIntOrNull()
            if (!response.isSuccessful || (code != null && code != 0)) {
                throw HermesLiveChatException(
                    error = mapBackendError(response.code, payload.optStringOrNull("code")),
                    backendCode = payload.optStringOrNull("code") ?: payload.optStringOrNull("error_code"),
                    message = payload.optStringOrNull("msg") ?: response.message,
                    status = response.code,
                )
            }
            if (payload.has("code") && payload.has("data")) {
                return@withContext payload.optJSONObject("data") ?: JSONObject()
            }
            payload
        }
    }

    private fun Request.Builder.applyHeaders(token: String?, hasBody: Boolean = false): Request.Builder {
        header("Accept", "application/json")
        if (hasBody) header("Content-Type", "application/json")
        if (token != null) header("Authorization", "Bearer $token")
        return this
    }
}

internal fun mapBackendError(status: Int, code: String?): HermesLiveChatError = when (code) {
    "70001" -> HermesLiveChatError.BAD_REQUEST
    "70002", "LC_TOKEN_INVALID" -> HermesLiveChatError.TOKEN_INVALID
    "70003", "LC_TOKEN_EXPIRED" -> HermesLiveChatError.TOKEN_EXPIRED
    "70004", "LC_INVALID_VISITOR_ID" -> HermesLiveChatError.INVALID_VISITOR_ID
    "70024", "LC_CONV_FORBIDDEN" -> HermesLiveChatError.CONVERSATION_FORBIDDEN
    "70025", "LC_CONV_CLOSED" -> HermesLiveChatError.CONVERSATION_CLOSED
    "LC_MESSAGE_RATE_LIMITED" -> HermesLiveChatError.MESSAGE_RATE_LIMITED
    "LC_CONTENT_INVALID" -> HermesLiveChatError.CONTENT_INVALID
    "70030", "LC_ATTACHMENT_TOO_LARGE" -> HermesLiveChatError.ATTACHMENT_TOO_LARGE
    "70031", "LC_ATTACHMENT_TYPE_INVALID", "LC_ATTACHMENT_TYPE_NOT_ALLOWED" -> HermesLiveChatError.ATTACHMENT_TYPE_INVALID
    "70011", "LC_CHANNEL_DISABLED" -> HermesLiveChatError.CHANNEL_DISABLED
    "70012", "LC_DOMAIN_NOT_ALLOWED" -> HermesLiveChatError.DOMAIN_NOT_ALLOWED
    "70010", "LC_ORG_LIVECHAT_DISABLED" -> HermesLiveChatError.ORG_DISABLED
    "70006", "LC_APP_INIT_TOKEN_INVALID" -> HermesLiveChatError.APP_INIT_TOKEN_INVALID
    "70007", "LC_APP_INIT_TOKEN_EXPIRED" -> HermesLiveChatError.APP_INIT_TOKEN_EXPIRED
    "LC_REALTIME_CONNECT_UNAUTHORIZED" -> HermesLiveChatError.REALTIME_CONNECT_UNAUTHORIZED
    "70050", "LC_REALTIME_PROVIDER_UNAVAILABLE" -> HermesLiveChatError.REALTIME_PROVIDER_UNAVAILABLE
    else -> when {
        status == 400 -> HermesLiveChatError.BAD_REQUEST
        status == 401 -> HermesLiveChatError.TOKEN_INVALID
        status == 403 -> HermesLiveChatError.CONVERSATION_FORBIDDEN
        status >= 500 -> HermesLiveChatError.REALTIME_PROVIDER_UNAVAILABLE
        else -> HermesLiveChatError.UNKNOWN
    }
}

File: ios/Sources/HermesLiveChat/Internal/ApiClient.swift
import Foundation

internal final class ApiClient {
    let config: HermesLiveChatConfig
    let session = URLSession(configuration: .default)

    init(config: HermesLiveChatConfig) {
        self.config = config
    }

    func publicConfig(locale: String?) async throws -> [String: Any] {
        try await get(
            path: "/api/livechat/v1/public-config",
            query: [
                "channel_type": "app",
                "app_key": config.appKey,
                "locale": locale,
            ].compactMapValues { $0 }
        )
    }

    func initSession(identity: VisitorIdentity, oldVisitorToken: String?) async throws -> [String: Any] {
        var body: [String: Any] = [
            "channel_type": "app",
            "app_key": config.appKey,
            "user": [
                "email": identity.email,
                "name": identity.name,
                "avatar": identity.avatar,
            ].compactMapValues { $0 },
        ]
        body["customer_id"] = identity.customerId
        body["external_user_id"] = identity.externalUserId
        body["business_id"] = identity.businessId
        body["ticket_id"] = identity.ticketId
        body["number"] = identity.number
        body["locale"] = identity.locale
        body["attrs"] = identity.attrs
        body["identity_token"] = identity.identityToken
        return try await post(path: "/api/livechat/v1/init", body: body.compactMapValues { $0 }, token: oldVisitorToken)
    }

    func sendText(token: String, conversationId: String?, text: String, clientMsgId: String) async throws -> SendMessageResult {
        var body: [String: Any] = [
            "client_msg_id": clientMsgId,
            "content_type": "text",
            "content": ["text": text],
        ]
        body["conversation_id"] = conversationId
        return SendMessageResult.from(try await post(path: "/api/livechat/v1/messages", body: body.compactMapValues { $0 }, token: token))
    }

    func sendImage(token: String, conversationId: String?, key: String, url: String, mimeType: String, size: Int, clientMsgId: String) async throws -> SendMessageResult {
        var body: [String: Any] = [
            "client_msg_id": clientMsgId,
            "content_type": "image",
            "content": ["key": key, "url": url, "mime": mimeType, "size": size],
        ]
        body["conversation_id"] = conversationId
        return SendMessageResult.from(try await post(path: "/api/livechat/v1/messages", body: body.compactMapValues { $0 }, token: token))
    }

    func markRead(token: String, messageId: String) async throws {
        _ = try await post(path: "/api/livechat/v1/messages/\(messageId.urlEncoded)/read", body: nil, token: token)
    }

    func history(token: String, conversationId: String, afterId: String?, limit: Int) async throws -> [Message] {
        let json = try await get(
            path: "/api/livechat/v1/conversations/\(conversationId.urlEncoded)/messages",
            query: [
                "limit": "\(limit)",
                "after_id": afterId,
            ].compactMapValues { $0 },
            token: token
        )
        return (json["items"] as? [[String: Any]] ?? []).map(Message.from)
    }

    func listConversations(token: String) async throws -> [Conversation] {
        let json = try await get(
            path: "/api/livechat/v1/conversations",
            query: ["limit": "20"],
            token: token
        )
        return (json["items"] as? [[String: Any]] ?? []).map(Conversation.from)
    }

    func presign(token: String, filename: String, mimeType: String, size: Int) async throws -> [String: Any] {
        try await post(
            path: "/api/livechat/v1/attachments/presign",
            body: ["filename": filename, "mime": mimeType, "size": size],
            token: token
        )
    }

    func uploadPresigned(url: URL, method: String, headers: [String: String], data: Data) async throws {
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.httpBody = data
        headers.forEach { request.setValue($0.value, forHTTPHeaderField: $0.key) }
        let (_, response) = try await session.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 500
        guard status < 300 else {
            throw HermesLiveChatException(error: .attachmentTypeInvalid, code: nil, message: "attachment upload failed", status: status)
        }
    }

    private func get(path: String, query: [String: String], token: String? = nil) async throws -> [String: Any] {
        var components = URLComponents(url: config.baseUrl.liveChatAppendingPath(path), resolvingAgainstBaseURL: false)!
        components.queryItems = query.map { URLQueryItem(name: $0.key, value: $0.value) }
        var request = URLRequest(url: components.url!)
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        return try await execute(request)
    }

    private func post(path: String, body: [String: Any]?, token: String?) async throws -> [String: Any] {
        var request = URLRequest(url: config.baseUrl.liveChatAppendingPath(path))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        if let token { request.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization") }
        return try await execute(request)
    }

    private func execute(_ request: URLRequest) async throws -> [String: Any] {
        let (data, response) = try await session.data(for: request)
        let status = (response as? HTTPURLResponse)?.statusCode ?? 500
        let payload = data.isEmpty ? [:] : (try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:])
        let code = payload["code"].flatMap { Int("\($0)") }
        if status < 200 || status >= 300 || (code != nil && code != 0) {
            throw HermesLiveChatException(
                error: mapBackendError(status: status, code: payload["code"].map { "\($0)" }),
                code: payload["code"].map { "\($0)" },
                message: payload["msg"] as? String,
                status: status
            )
        }
        if payload.keys.contains("code"), let data = payload["data"] as? [String: Any] {
            return data
        }
        return payload
    }
}
